core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala

```diff
@@ -192,12 +193,16 @@ class ExternalAppendOnlyMap[K, V, C](
    * It will be called by TaskMemoryManager when there is not enough memory for the task.
    */
   override protected[this] def forceSpill(): Boolean = {
-    assert(readingIterator != null)
-    val isSpilled = readingIterator.spill()
-    if (isSpilled) {
-      currentMap = null
+    if (isReadingIterator) {
+      assert(readingIterator != null)
+      val isSpilled = readingIterator.spill()
+      if (isSpilled) {
+        currentMap = null
+      }
+      isSpilled
+    } else {
+      false
```

Comment:
The reason I chose to simple return false is, if the in-memory map is not yet spilled to disk during inserting data, it may not need to be spilled in fact as it should be a small spill.

But I think it is arguable that when `forceSpill` is called, we should spill the collection anyway. I will do the change.